File: services/CacheService.rb
#
# Check local cache, if not present, fetch from web and save to json files
#

require "http"

class CacheService
    def cachePositions(force)
        Logger.debug "Cache positions"

        @filename = Filenames.positions()
        @endpoint = "/positions"

        deleteCacheIfNeeded(force)
        checkCacheAndSave()
    end

    def cacheTeams(force)
        Logger.debug "Cache teams"

        @filename = Filenames.teams()
        @endpoint = "/teams"

        deleteCacheIfNeeded(force)
        checkCacheAndSave()
    end

    def cacheTeamRoster(team, force)
        Logger.debug "Cache roster for team #{team.name}"

        @filename = Filenames.teamRoster(team)
        @endpoint = "/roster/#{team.abbreviation}/#{Constants.currentSeason}"

        deleteCacheIfNeeded(force)
        checkCacheAndSave()
    end

    def cachePlayerForId(id, force)
        Logger.debug "Cache player with ID #{id}"

        @filename = Filenames.playerForId(id)
        @endpoint = "/player/#{id}/landing"

        deleteCacheIfNeeded(force)
        checkCacheAndSave()
    end

    private

    def filePath
        return "./json/#{@filename}"
    end

    def getCachedData
        content = File.read(filePath)
        return JSON.parse(content)
    end

    def apiEndpoint
        return "https://api-web.nhle.com/v1" + @endpoint
    end

    def deleteCacheIfNeeded(force)
        if force && File.exist?(filePath)
            Logger.info "Deleting file at path #{filePath}"
            File.delete(filePath)
        end
    end

    def checkCacheAndSave
        if File.exists?(filePath)
            Logger.debug "Cache file exist, nothing to do"
        else
            Logger.warning "Calling NHL's API"
            Logger.debug apiEndpoint
            response = HTTP.get(apiEndpoint)
            File.write(filePath, response)
        end
    end
end

File: src/main/kotlin/ca/sebleclerc/hockeydata/services/CacheService.kt
package ca.sebleclerc.hockeydata.services

import ca.sebleclerc.hockeydata.helpers.Logger
import ca.sebleclerc.hockeydata.models.CacheStep
import okhttp3.OkHttpClient
import okhttp3.Request

class CacheService(private val import: ImportService) {
  private val apiClient = OkHttpClient()

  fun cache(steps: List<CacheStep>, force: Boolean) {
    steps.forEach { cacheStep(it, force) }

    import.importRosters()
  }

  private fun cacheStep(step: CacheStep, force: Boolean) {
    deleteCacheIfNeeded(step, force)
    checkCacheAndSave(step)
  }

  private fun deleteCacheIfNeeded(step: CacheStep, force: Boolean) {
    if (force && step.file.exists()) {
      Logger.info("Deleting file at path.")
      step.file.delete()
    }
  }

  private fun checkCacheAndSave(step: CacheStep) {
    if (step.file.exists()) {
      Logger.info("Cache file exist, nothing to do.")
    } else {
      Logger.warning("Calling NHL's API")
      Logger.info(step.apiPath)
      val request = Request
        .Builder()
        .url(step.apiPath)
        .build()

      val response = apiClient.newCall(request).execute()
      val body = response.body?.string() ?: ""
      step.file.writeText(body)
    }
  }
}
